File: app/src/test/java/com/pietrantuono/podcasts/addpodcast/singlepodcast/model/PodcastModelImplTest.java
package com.pietrantuono.podcasts.addpodcast.singlepodcast.model;

import com.pietrantuono.podcasts.addpodcast.repository.repository.Repository;
import com.pietrantuono.podcasts.apis.SinglePodcastApi;
import com.pietrantuono.podcasts.addpodcast.model.pojos.Podcast;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.observers.TestObserver;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class PodcastModelImplTest {
    private static final String TEXT = "text";
    private static final int TRACK_ID = 1;
    @Mock SinglePodcastApi api;
    @Mock Repository repository;
    @Mock Podcast podcast;
    private SinglePodcastModelImpl model;

    @Before
    public void setUp() throws Exception {
        model = new SinglePodcastModelImpl(api, repository);
        when(podcast.getTrackId()).thenReturn(TRACK_ID);
        when(podcast.getFeedUrl()).thenReturn(TEXT);
        when(repository.getIfSubscribed(TRACK_ID)).thenReturn(Observable.just(true));
        model.startModel(podcast);
    }

    @Test
    public void when_getIsSubscribed_then_getsIfSubscribed() {
        //THEN
        verify(repository).getIfSubscribed(TRACK_ID);
    }

    @Test
    public void when_subscribesToIsSubscribed_then_subscribes() {
        //WHEN
        TestObserver<Boolean> ifSubscribedObserver = new TestObserver<>();
        model.subscribeToIsSubscribedToPodcast(ifSubscribedObserver);
        //THEN
        List<Boolean> result = new ArrayList<>();
        result.add(true);
        ifSubscribedObserver.assertReceivedOnNext(result);
    }

    @Test
    public void when_init_then_getsFeed() {
        //THEN
        verify(api).getFeed(TEXT);
    }

    @Test
    public void when_init_then_getsIsSubscribedToPodcast() {
        //THEN
        verify(repository).getIfSubscribed(TRACK_ID);
    }

    @Test
    public void when_actuallySubscribes_then_repositorySubscribes() {
        //WHEN
        model.actuallySubscribesToPodcast();
        //THEN
        verify(repository).actuallySubscribesToPodcast(podcast);
    }

    @Test
    public void when_actuallyUnSubscribes_then_repositoryUnSubscribes() {
        //WHEN
        model.actuallyUnSubscribesToPodcast();
        //THEN
        verify(repository).actuallyUnSubscribesToPodcast(podcast);
    }
}

File: app/src/test/java/com/pietrantuono/podcasts/addpodcast/singlepodcast/model/PodcastModelImplTest.kt
package com.pietrantuono.podcasts.addpodcast.singlepodcast.model

import com.pietrantuono.podcasts.addpodcast.model.pojos.Podcast
import com.pietrantuono.podcasts.apis.PodcastFeed
import com.pietrantuono.podcasts.apis.SinglePodcastApi
import com.pietrantuono.podcasts.repository.repository.PodcastRepo
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import rx.Observable
import rx.observers.TestObserver
import rx.schedulers.Schedulers
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class PodcastModelImplTest {
    @Mock lateinit var api: SinglePodcastApi
    @Mock lateinit var repository: PodcastRepo
    @Mock lateinit var podcast: Podcast
    lateinit var model: SinglePodcastModelImpl
    @Mock lateinit var feed: PodcastFeed
    private val io = Schedulers.trampoline()
    private val main = Schedulers.trampoline()

    @Before
    @Throws(Exception::class)
    fun setUp() {
        model = SinglePodcastModelImpl(api, repository, io, main)
        `when`<String>(podcast.feedUrl).thenReturn(TEXT)
        `when`(repository.getIfSubscribed(podcast)).thenReturn(Observable.just<Boolean>(true))
        val observable = Observable.just(feed)
        `when`(api.getFeed(ArgumentMatchers.anyString())).thenReturn(observable)
        model.startModel(podcast)
    }


    @Test
    fun when_subscribesToIsSubscribed_then_subscribes() {
        //WHEN
        val ifSubscribedObserver = TestObserver<Boolean>()
        model.subscribeToIsSubscribedToPodcast(ifSubscribedObserver)
        //THEN
        val result = ArrayList<Boolean>()
        result.add(true)
        ifSubscribedObserver.assertReceivedOnNext(result)
    }

    @Test
    fun when_init_then_getsFeed() {
        //THEN
        verify<SinglePodcastApi>(api).getFeed(TEXT)
    }

    companion object {
        private val TEXT = "text"
        private val TRACK_ID = 1
    }
}